### PR TITLE
Make sure foreground access works for DefTerm

### LINKED ISSUE #13211: GUI apps still don't open in the foreground, from a default terminal launch
A TCC user said that GUI apps started by TCC in WT don't get focus (unlike other shells in WT and unlike TCC in a console).  I said I didn't think I could get TCC and/or WT to act that way if I tried.  Have I overlooked anything ... any ideas?  Thanks!

### PR DESCRIPTION
**Pull request by zadjii-msft**
See also: #12799, the origin of much of this.

This change evolved over multiple phases. 

### Part the first

When we create a defterm connection in `TerminalPage::_OnNewConnection`,
we don't have the hosting HWND yet, so the tab gets created without one.
We'll later get called with the owner, in `Initialize`. 

To remedy this, we need to:
* In `Initialize`, make sure to update any existing controls with the
  new owner.
* In `ControlCore`, actually propogate the new owner down to the
  connection

### Part the second

DefTerm launches don't actually request focus mode, so the Terminal
never sends them focus events. We need those focus events so that the
console can request foreground rights.

To remedy this, we need to:
* pass `--win32input` to the commandline used to initialize OpenConsole
  in ConPTY mode. We request focus events at the same time we request
  win32-input-mode.
* I also added `--resizeQuirk`, because _by all accounts that should be
  there_. Resizing in defterm windows should be _wacky_ without it, and
  I'm a little surprised we haven't seen any bugs due to this yet.

### Part the third

`ConsoleSetForeground` expects a `HANDLE` to the process we want to give
foreground rights to. The problem is, the wire format we used _also_
decided that a HANDLE value was a good idea. It's not. If we pass the
literal value of the HANDLE to the process from OpenConsole to conhost,
so conhost can call that API, the value that conhost uses there will
most likely be an invalid handle. The HANDLE's value is its value in
_OpenConsole_, not in conhost.

To remedy this, we need to:
* Just not forward `ConsoleSetForeground`. Turns out, we _can_ just call
  that in OpenConsole safely. There's no validation. So just instantiate
  a static version of the Win32 version of ConsoleControl, just to use
  for SetForeground. (thanks Dustin)

* [x] Tested manually - Win+R `powershell`, `notepad` spawns on top.

Closes #13211

### CODE REVIEW

File: src/server/ProcessHandle.cpp [hunk 1]
```diff
@@ -34,6 +37,44 @@ ConsoleProcessHandle::ConsoleProcessHandle(const DWORD dwProcessId,
...
+    const auto& conhost{ ServiceLocator::LocateGlobals().handoffInboxConsoleHandle };
+    if (_hProcessInConhost && conhost)
+    {
+        LOG_IF_WIN32_BOOL_FALSE(DuplicateHandle(conhost.get(),
+                                                _hProcessInConhost.get(),
+                                                nullptr /* hTargetProcessHandle */,
+                                                nullptr /* lpTargetHandle, ignored */,
+                                                0 /* dwDesiredAccess, ignored */,
+                                                false,
+                                                DUPLICATE_CLOSE_SOURCE));
```
Comment (DHowett) [inline src/server/ProcessHandle.cpp:76]: HAAHAHAHAHA OH GOD


I read your description and was all like, "Are we about to cause a handle leak in conhost"

Here I see that my concerns were unfounded

Comment (zadjii-msft) [reply src/server/ProcessHandle.cpp:76]: nah see I'm a good programmer (on Wednesdays) 😄

Comment (zadjii-msft) [reply src/server/ProcessHandle.cpp]: lmao I was a bad programmer that Wednesday 🙃

File: src/inc/HostSignals.hpp [hunk 2]
```diff
@@ -13,21 +13,21 @@ namespace Microsoft::Console
...
         uint32_t sizeInBytes;
-        uint32_t processId;
+        uint32_t processId; // THIS IS A PID
     };
 
     struct HostSignalSetForegroundData
     {
         uint32_t sizeInBytes;
-        uint32_t processId;
+        uint32_t processId; // THIS IS A HANDLE, NOT A PID
```
Comment (DHowett) [inline src/inc/HostSignals.hpp:22]: blame @miniksa

**Comment by DHowett** [comment]
Nits: When you prep this as a commit message, remove the commit IDs from the body (they aren't meaningful once squashed), and remove the original notes about the thing we didn't do / the summary/details section

**Comment by DHowett** [comment]
Moved from body:

----

See also: #12799, the origin of much of this.

This change evolved over multiple phases. 

### Part the first (4de1ef73e9a8e0fb7582d60c6e1152fb9aeb1a77)

When we create a defterm connection in `TerminalPage::_OnNewConnection`, we don't have the hosting HWND yet, so the tab gets created without one. We'll later get called with the owner, in `Initialize`. 

To remedy this, we need to:
* In `Initialize`, make sure to update any existing controls with the new owner.
* In `ControlCore`, actually propogate the new owner down to the connection

### Part the second (5ab079915b27438850d516e521960d71504b78d2)

DefTerm launches don't actually request focus mode, so the Terminal never sends them focus events. We need those focus events so that the console can request foreground rights.

To remedy this, we need to:
* pass `--win32input` to the commandline used to initialize OpenConsole in ConPTY mode. We request focus events at the same time we request win32-input-mode.
* I also added `--resizeQuirk`, because _by all accounts that should be there_. Resizing in defterm windows should be _wacky_ without it, and I'm a little surprised we haven't seen any bugs due to this yet.


### ~Part the third (e970f7c93c4005d8a138d4f2e6acab965fe43831)~

<details>
<summary>Part the third was changed dramatically, original summary here.</summary>

There's certain OS-side checks that make it so only the original `conhost` can call the `ConsoleControl` APIs. When launched as the delegation console, OpenConsole forwards these calls to the original conhost that spawned it. The one in question here is `ConsoleSetForeground`. That expects a `HANDLE` to the process we want to give foreground rights to. The problem is, the wire format we used _also_ decided that a HANDLE value was a good idea. It's not. If we pass the literal value of the HANDLE to the process from OpenConsole to conhost, so conhost can call that API, the value that conhost uses there will most likely be an invalid handle. The HANDLE's value is its value in _OpenConsole_, not in conhost.

To remedy this, we need to:
* Duplicate client process handles to the original conhost when running in delegation console mode
* Pass the value of the handle in the conhost process space to RemoteConsoleControl instead.

This is necessarily tricky, because we assumingly can't make conhost-side changes here. We have to assume that new Terminal's will be running on older versions of Windows 11, and also older Terminals will be running on newer Windows 11's (with this fix). If there's a mismatch in what either is expecting, then the whole thing'll blow up. 

By duplicating the handle into the conhost, we can keep passing a HANDLE value that conhost will pass straight through on all versions of Windows 11. That also means OpenConsole needs to manage the lifetime of that handle, because old conhosts will NEVER be able to know that there's a new HANDLE in their process space. 

</details>

### Part the fourth, (55a7331)

`ConsoleSetForeground` expects a `HANDLE` to the process we want to give foreground rights to. The problem is, the wire format we used _also_ decided that a HANDLE value was a good idea. It's not. If we pass the literal value of the HANDLE to the process from OpenConsole to conhost, so conhost can call that API, the value that conhost uses there will most likely be an invalid handle. The HANDLE's value is its value in _OpenConsole_, not in conhost.

To remedy this, we need to:
* Just not forward `ConsoleSetForeground`. Turns out, we _can_ just call that in OpenConsole safely. There's no validation. So just instantiate a static version of the Win32 version of ConsoleControl, just to use for SetForeground. (thanks Dustin)

* [x] Closes #13211
* [x] Tested manually - Win+R `powershell`, `notepad` spawns on top.
* [x] I work here

**Review by lhecker — APPROVED** [review]
I love this PR. The comments are great and explain the problems well.
For an outsider understand what `--resizeQuirk` and `--win32input` _mean_ is not really explained in this PR and might be hard to understand, but I believe that this is clearly out of scope.

File: src/inc/HostSignals.hpp (hunk 2 above)
Comment (miniksa) [reply src/inc/HostSignals.hpp:22]: Brain fart. Sorry.

**Pull request merged**